connector.js

```diff
@@ -9,125 +11,129 @@
+const zlib = require('zlib');
 
 class HttpAmazonESConnector extends HttpConnector {
   constructor(host, config) {
     super(host, config);
-    const { protocol, port } = host;
+
+    const protocol = host.protocol;
+    const port = host.port;
     const endpoint = new AWS.Endpoint(host.host);
 
     if (protocol) endpoint.protocol = protocol.replace(/:?$/, ":");
     if (port) endpoint.port = port;
 
-    this.AWS = AWS;
     this.awsConfig = config.awsConfig || AWS.config;
     this.endpoint = endpoint;
     this.httpOptions = config.httpOptions || this.awsConfig.httpOptions;
+    this.httpClient = new AWS.NodeHttpClient();
   }
 
-  async request(params, cb) {
-    let incoming;
-    let timeoutId;
-    let request;
+  request(params, cb) {
+    const reqParams = this.makeReqParams(params);
+
     let req;
     let status = 0;
     let headers = {};
-    let log = this.log;
     let response;
-    const AWS = this.AWS;
+    let incoming;
+    let cancelled;
+
+    const cancel = () => {
+      cancelled = true;
+      req && req.abort();
+    };
 
-    let reqParams = this.makeReqParams(params);
     // general clean-up procedure to run after the request
     // completes, has an error, or is aborted.
-    let cleanUp = _.bind(function (err) {
-      clearTimeout(timeoutId);
-
+    const cleanUp = (err) => {
       req && req.removeAllListeners();
       incoming && incoming.removeAllListeners();
 
-      if ((err instanceof Error) === false) {
-        err = void 0;
-      }
-
-      log.trace(params.method, reqParams, params.body, response, status);
-      if (err) {
+      this.log.trace(params.method, reqParams, params.body, response, status);
+      if (err instanceof Error) {
         cb(err);
       } else {
-        cb(err, response, status, headers);
+        cb(null, response, status, headers);
       }
-    }, this);
+    };
+
+    // load creds
+    this.getAWSCredentials()
+      .catch(e => {
+        if (e && e.message) e.message = `AWS Credentials error: ${e.message}`;
+        throw e;
+      })
+      .then(creds => {
+        if (cancelled) {
+          return;
+        }
+
+        const request = this.createRequest(params, reqParams);
+
+        // Sign the request (Sigv4)
+        this.signRequest(request, creds);
+
+        req = this.httpClient.handleRequest(request, this.httpOptions, function (_incoming) {
+          incoming = _incoming;
+          status = incoming.statusCode;
+          headers = incoming.headers;
+          response = '';
+
+          let encoding = (headers['content-encoding'] || '').toLowerCase();
+          if (encoding === 'gzip' || encoding === 'deflate') {
+            incoming = incoming.pipe(zlib.createUnzip());
+          }
+
+          incoming.setEncoding('utf8');
+          incoming.on('data', function (d) {
+            response += d;
+          });
+
+          incoming.on('error', cleanUp);
+          incoming.on('end', cleanUp);
+        }, cleanUp);
```

Comment:
I removed `req.on('error', cleanUp)` -- the tests I wrote showed that the AWS HttpClient already attaches an .on('error') listener to the req, and calls the errCallback, so this actually resulted in cleanUp (and thus the log message, request callback, etc) being called twice!